src/wp-includes/block-template-utils.php

```diff
@@ -37,11 +37,9 @@
  * }
  */
 function get_block_theme_folders( $theme_stylesheet = null ) {
-	$theme_name = null === $theme_stylesheet ? get_stylesheet() : $theme_stylesheet;
-	$root_dir   = get_theme_root( $theme_name );
-	$theme_dir  = "$root_dir/$theme_name";
+	$theme = wp_get_theme( $theme_stylesheet );
 
-	if ( file_exists( $theme_dir . '/block-templates' ) || file_exists( $theme_dir . '/block-template-parts' ) ) {
+	if ( $theme->has_block_templates() ) {
```

Comment:
Instead of caching in `WP_Theme` whether a theme has the old `block-templates` / `block-template-parts` folders, why not cache the actual folder names in `WP_Theme`? I think that'll encapsulate the necessary functionality even more efficiently by avoiding the if clause every time this function is called.

I'm thinking we could introduce a `WP_Theme::get_block_template_folders()` or `WP_Theme::get_template_folders()` method instead of `WP_Theme::has_block_templates()`.

That would simplify this method even more, basically just to `return $theme->get_block_template_folders();`.

Reply:
The reason we are storing this data is the file look is expensive. The if statement after that is cheap and I am not worried about caching. Caching brings up it's own problems. What if we want to add / update / delete this cache. I would do that cache invalidation. 

One other nice thing we can do this in change, is populate the cache ( the `has_block_templates` property ) from `is_block_theme`.

Reply:
Right, but why not cache the value we actually care about? Caching a specific piece of logic just because it's expensive is good enough for performance, but it's not really future-proof: What if the logic to get the folder names changes in the future? Caching the folder names in a property will allow us to change the logic at any point without deprecating anything, while caching a piece of the logic _currently_ needed isn't as future-proof.

Additionally, the `has_block_templates` name is confusing as it is not about whether the theme has block templates, but whether the theme uses the old naming convention for block theme folders.

Reply:
@felixarntz I agree with you here. We have addressed this in https://github.com/WordPress/wordpress-develop/pull/5202